scripts/redcap/wrong_date_associations.py

```diff
@@ -0,0 +1,280 @@
+                                  'parse_dates': fields,
+                              })
+    return data
+
+
+def mark_lagging_dates(data, comparison_var, days_duration):
+    comparisons = data.loc[:, [comparison_var]]
+    df = data.drop(columns=[comparison_var]).copy()
+    df.columns.name = 'form_date_var'  # so that stacking index has a name
+    data_long = df.stack().to_frame('date').reset_index(-1)
+    data_comp = data_long.join(comparisons)
+    # Maybe extract previous code into data prep?
+    data_comp['precedes'] = data_comp['date'] < data_comp['visit_date']
+    data_comp['exceeds'] = data_comp['date'] > (
+        data_comp['visit_date'] + pd.Timedelta(days=days_duration))
+    data_comp['purgable'] = data_comp['precedes'] | data_comp['exceeds']
+    return data_comp
+
+
+def log_dataframe_by_row(errors_df: pd.DataFrame,
+                         uid_template: str = "{id}/{redcap_event_name}/{form}",
+                         **kwargs):
+    """
+    Convert each row of the DataFrame into a Sibislogger issue.
+    General idea: Each column in errors_df is reported, each additional keyword
+    argument is a template that gets populated with data from the columns.
+
+    Note: imported from 'error_handling.py' within 'hivalc-data-integration'
+    """
+
+    def log_row(row: pd.Series, **kwargs):
+        kwargs.update({
+            k: v.format(**row) for k, v in kwargs.items()
+            if isinstance(v, string_types)
+        })
+        slog.info(
+            **kwargs,
+            **row.dropna())
+
+    for _, row in errors_df.reset_index().iterrows():
+        log_row(row, uid=uid_template, **kwargs)
+
+def subtract_special_cases_from_marks(marks, session):
+    """
+    Looks at special_cases.yml file, finds the special cases under the file name section,
+    and then excludes any cases that match the special cases list.
+    
+    Format of Special Case:
+     - subject: A-00022-F-1
+       event: 1y_visit_arm_1
+       dates: bio_np_date
+
+    Notes:
+
+    For fields used within Pandas dataframe:
+    - 'event_name' is REDCap event and arm name
+    - 'study_id' inclues up to letter and number
+    - 'form_date_var' is comparison date variable
+
+    If using DataFram.iterrows(), then Each row in the dataframe is a pd.Series which has:
+    row.name -> [0] is the subject ID, [1] is the arm name
+    """
+
+    # Connect to the special_cases.yml file
+    sibis_config = session.get_operations_dir()
+    special_cases_file = os.path.join(sibis_config, 'special_cases.yml')
+    if not os.path.isfile(special_cases_file):
+        slog.info("Special cases file does not exit","Error: The following file does not exist : " + special_cases_file)
+        sys.exit(1)
+
+    # Get a list of the specific cases that should be inspected
+    with open(special_cases_file) as fi:
+        special_cases = yaml.load(fi)
+        exceptions_data = special_cases.get('wrong_date_associations', [])
```

Comment:
To get the triplets that allow you to select by pandas index, and to facilitate `dates` that could be either list or string, this should work:

```python
from typing import Sequence, List, Dict, Tuple

def create_special_cases_triplets(exceptions_data: List[Dict]) -> List[Tuple[str]]:
    cases = []
    for case in exceptions_data:
        if isinstance(case['dates'], Sequence) and not isinstance(case['dates'], str):
            cases.extend([(case['subject'], case['event'], x) for x in case['dates']])
        elif isinstance(case['dates'], str):
            cases.append((case['subject'], case['event'], case['dates']))
        else:
            # log error?
            pass

    return cases
```